src/mono/mono/mini/mini-runtime.c

```diff
@@ -5395,3 +5401,33 @@ mini_alloc_jinfo (MonoJitMemoryManager *jit_mm, int size)
 		return (MonoJitInfo*)mono_mem_manager_alloc0 (jit_mm->mem_manager, size);
 	}
 }
+
+void
+mono_set_runtime_init_callback (MonoRuntimeInitCallback callback)
+{
+	runtime_init_callback = callback;
+}
+
+/*
+* Default implementation invoking runtime init callback in native-to-managed wrapper.
+*/
+void
+mono_invoke_runtime_init_callback (void)
```

Comment:
I don't understand why I would ever want to pass something other than this method to `runtime_init_callback=`

> It is also possible to pass runtime-init-callback= to AOT compiler, and in that case native-to-managed wrapper will call that function and its up to implementor to do a thread safe implementation of lazy runtime init. This capability could be used in case where the library can't set the callback before consumer of the library calls the exported function.

So a library user is supposed to call `mono_set_runtime_init_callback` and pass it... what... something that calls `mono_jit_init_version()` ?

And the issue is that maybe they have a race between doing that and calling the first n2m method? And so we give them an escape hatch by letting them also specify something for `runtime_init_callback=` to do their atomic waiting for them?

If they can pass something arbitrary to `runtime_init_callback` why do they need `mono_set_runtime_init_callback` at all?

Reply:
So the background of the complete feature is ability to compile runtime and managed code into a self contained shared/static library, similar to what NativeAOT does in their library mode and have a lazy initi of runtime on first call to exported entrypoint. In the case when building a shared library, library builder needs to have a way to run custom init code of the runtime, so it needs to either supply a callback or compile in a specific call to a function very early in the native-to-managed wrapper that does all needed runtime initialization. Since consumer of the library will just do dlopen/dlsym + call entrypoint (really a native-to-managed wrapper), the library build needs to have code running on dlopen in order to set any callback that gets called. That can be achieved by using C++ global constructor or things like `__attribute__((constructor))`, but if none of them are options, the only way to get control is to compile in a custom function call when building the complete library. So I opened up to support both scenarios.

The first is when the caller has the ability to set callback early using some mechanism that is supported by compiler (C++ global constructor, `__attribute__((constructor))` or pragma and compiler/linker specific features), but if there is no support for that, meaning the embedder can't set the callback early enough, then there is the ability to supply a symbol at compile time compiled into the native-to-managed wrapper.

The `mono_set_runtime_init_callback/mono_invoke_runtime_init_callback` implementation of the that is very close to how NativeAOT does it. Here is where they set the callback using a global C++ constructor:

https://github.com/dotnet/runtime/blob/b01311bd8dc591ce26b88a86b592c63a26798ceb/src/coreclr/nativeaot/Bootstrap/main.cpp#L227

and here is the implementation in their reverse pinvoke wrapper (similar to our native-to-managed wrapper):

https://github.com/dotnet/runtime/blob/50f9bcc663c552d8658133c707332778a000b7ef/src/coreclr/nativeaot/Runtime/thread.cpp#L1184

and here is the implementation of their corresponding mono_invoke_runtime_init_callback function:

https://github.com/dotnet/runtime/blob/50f9bcc663c552d8658133c707332778a000b7ef/src/coreclr/nativeaot/Runtime/thread.cpp#L1215

We could go different ways here (I think I considered most of them before landing on current implementation):

* Support things like done in this PR.
* Drop and just keep one of the mechanism, if we only go with one of the approaches.
* Piggy back on our call to mono_threads_attach_coop and implement the call to the callback early in mono_threads_attach_coop, then we won't need changes to native-to-managed wrapper, but we need to special case call to mono_threads_attach_coop a little since it must end up as a direct icall in the end.

The callback should do all runtime init normally done by the embedder, so register aot modules (if used), register bundles (if used), configure debugger (if used), calling mono_jit_init_version etc.

Reply:
> The callback should do all runtime init normally done by the embedder, so register aot modules (if used), register bundles (if used), configure debugger (if used), calling mono_jit_init_version etc. It can be customized by library builder depending on what features the build library should support.

I see, so the callback really has to be customized not by the end-user but by our task that builds and bundles the runtime and the user's code.

And so the expectation is that the task will decide based on some setting (or configuration, or platform, etc) whether it can insert an appropriate `__attribute__((constructor))` that will call to install the initialization callback, or else it will pass a flag to the AOT compiler to hardcode a symbol name.

That makes sense, thanks.

PS: I don't like using thread_attach for this.

PPS: I like that the synchronization logic is in the runtime. That means we could potentially inline the fast path (no callback, or callback was already called) into the wrapper.

Reply:
OK, then I will keep the support implemented in this PR.